prow/spyglass/gcsartifact_fetcher.go

```diff
@@ -152,81 +154,88 @@ const (
+func (af *GCSArtifactFetcher) signURL(ctx context.Context, linkPrefix, bucket, obj string) (string, error) {
+	if linkPrefix == providers.GS+"://" {
+		// We specifically want to use cookie auth, see:
+		// https://cloud.google.com/storage/docs/access-control/cookie-based-authentication
+		if af.useCookieAuth {
+			artifactLink := &url.URL{
+				Scheme: httpsScheme,
+				Host:   cookieHost,
+				Path:   path.Join(bucket, obj),
+			}
+			return artifactLink.String(), nil
 		}
-		return artifactLink.String(), nil
-	}
 
-	// If we're anonymous we can just return a plain URL.
-	if af.gcsCredsFile == "" {
-		artifactLink := &url.URL{
-			Scheme: httpsScheme,
-			Host:   anonHost,
-			Path:   path.Join(bucket, obj),
+		// If we're anonymous we can just return a plain URL.
+		if af.gcsCredsFile == "" {
+			artifactLink := &url.URL{
+				Scheme: httpsScheme,
+				Host:   anonHost,
+				Path:   path.Join(bucket, obj),
+			}
+			return artifactLink.String(), nil
 		}
-		return artifactLink.String(), nil
-	}
 
-	// TODO(fejta): do not require the json file https://github.com/kubernetes/test-infra/issues/16489
-	// As far as I can tell, there is no sane way to get these values other than just
-	// reading them out of the JSON file ourselves.
-	f, err := os.Open(af.gcsCredsFile)
-	if err != nil {
-		return "", err
-	}
-	defer f.Close()
-	auth := struct {
-		Type        string `json:"type"`
-		PrivateKey  string `json:"private_key"`
-		ClientEmail string `json:"client_email"`
-	}{}
-	if err := json.NewDecoder(f).Decode(&auth); err != nil {
-		return "", err
-	}
-	if auth.Type != "service_account" {
-		return "", fmt.Errorf("only service_account GCS auth is supported, got %q", auth.Type)
+		// TODO(fejta): do not require the json file https://github.com/kubernetes/test-infra/issues/16489
+		// As far as I can tell, there is no sane way to get these values other than just
+		// reading them out of the JSON file ourselves.
+		f, err := os.Open(af.gcsCredsFile)
+		if err != nil {
+			return "", err
+		}
+		defer f.Close()
+		auth := struct {
+			Type        string `json:"type"`
+			PrivateKey  string `json:"private_key"`
+			ClientEmail string `json:"client_email"`
+		}{}
+		if err := json.NewDecoder(f).Decode(&auth); err != nil {
+			return "", err
+		}
+		if auth.Type != "service_account" {
+			return "", fmt.Errorf("only service_account GCS auth is supported, got %q", auth.Type)
+		}
+		return storage.SignedURL(bucket, obj, &storage.SignedURLOptions{
+			Method:         "GET",
+			Expires:        time.Now().Add(10 * time.Minute),
+			GoogleAccessID: auth.ClientEmail,
+			PrivateKey:     []byte(auth.PrivateKey),
+		})
 	}
-	return storage.SignedURL(bucket, obj, &storage.SignedURLOptions{
-		Method:         "GET",
-		Expires:        time.Now().Add(10 * time.Minute),
-		GoogleAccessID: auth.ClientEmail,
-		PrivateKey:     []byte(auth.PrivateKey),
-	})
+	return af.opener.SignedURL(ctx, fmt.Sprintf("%s%s/%s", linkPrefix, bucket, obj))
```

Comment:
I think the GCS implementation of this should also be moved into pkio or not? Because of this we have to pass around both an `io.Opener` and a `gcsCredsFile`

Reply:
Since that will require adding more options to `pkio.NewOpener` it would probably be nice to use the functional Opts pattern there

Reply:
I can move it there. We just had a discussion a few PRs ago where the idea was to not put special GCS handling inside the opener package. But of course we can reconsider now that we see the problem which wasn't obivous at that point :)

Reply:
yeah, basically we have to decide between making the interface a bit leaky (passing in provider-specific options) and making it incomplete. Since the latter means that ppl in the future might think they can use it when they actually can not I think the former is the lesser evil. Additionally, it means we properly separate concerns.

Reply:
I moved it into the opener package. I think it's better to add the option to `SignedUrl` than to `NewOpener` as it's only used for `SignedUrl`. I'm not sure which options pattern to use. For now I just added `SignedURLOptions` wit a non-pointer flag for `UseGSCookieAuth`.

Just not sure if I should change it because it's inconsistent:
* we used a vararg in the Writer method instead (because we wanted to stay compatible)
* we used pointer vars instead of regular types

But I think when ignoring inconsistencies it's better to implement it this way, because we can avoid using pointers and it would be more complicated (and I think it doesn't make sense) to handle a vararg of `SignedURLOptions`. WDYT?

Reply:
> Wouldn't it make more sense to pass this in when creating the opener? As far as I can see we either do this always or never or did I miss something?

Yup, I can move it the NewOpener. My only concern was that adding UseGSCookieAuth to NewOpener would suggest CookieAuth is used for the GCS client (for reader, writer,..). But I guess this could be fixed by naming it something like UseGSCookieAuthForSignedURLs

WDYT?